shared/src/types/key/secp256k1.rs

```diff
@@ -0,0 +1,488 @@
+        >,
+    ) {
+        // Encoded as `[u8; COMPRESSED_PUBLIC_KEY_SIZE]`
+        let elements = "u8".into();
+        let length = libsecp256k1::util::COMPRESSED_PUBLIC_KEY_SIZE as u32;
+        let definition = borsh::schema::Definition::Array { elements, length };
+        definitions.insert(Self::declaration(), definition);
+    }
+
+    fn declaration() -> borsh::schema::Declaration {
+        "secp256k1::PublicKey".into()
+    }
+}
+
+#[allow(clippy::derive_hash_xor_eq)]
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.serialize_compressed().hash(state);
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.serialize_compressed().partial_cmp(&other.0.serialize_compressed())
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.serialize_compressed().cmp(&other.0.serialize_compressed())
+    }
+}
+
+impl Display for PublicKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", hex::encode(&self.0.serialize_compressed()))
+    }
+}
+
+impl FromStr for PublicKey {
+    type Err = ParsePublicKeyError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let vec = hex::decode(s).map_err(ParsePublicKeyError::InvalidHex)?;
+        BorshDeserialize::try_from_slice(&vec)
+            .map_err(ParsePublicKeyError::InvalidEncoding)
+    }
+}
+
+impl From<libsecp256k1::PublicKey> for PublicKey {
+    fn from(pk: libsecp256k1::PublicKey) -> Self {
+        Self(pk)
+    }
+}
+
+/// Secp256k1 secret key
+#[derive(Debug, Clone)]
+pub struct SecretKey(pub libsecp256k1::SecretKey);
+
+impl super::SecretKey for SecretKey {
+    type PublicKey = PublicKey;
+
+    const TYPE: SchemeType = SigScheme::TYPE;
+
+    fn try_from_sk<PK: super::SecretKey>(
+        pk: &PK,
+    ) -> Result<Self, ParseSecretKeyError> {
+        if PK::TYPE == super::common::SecretKey::TYPE {
+            super::common::SecretKey::try_from_sk(pk).and_then(|x| match x {
+                super::common::SecretKey::Secp256k1(epk) => Ok(epk),
+                _ => Err(ParseSecretKeyError::MismatchedScheme),
+            })
+        } else if PK::TYPE == Self::TYPE {
+            Self::try_from_slice(pk.try_to_vec().unwrap().as_slice())
+                .map_err(ParseSecretKeyError::InvalidEncoding)
+        } else {
+            Err(ParseSecretKeyError::MismatchedScheme)
+        }
+    }
+}
+
+impl Serialize for SecretKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> 
+    where
+        S: Serializer,
+    {
+        // not sure if this is how I should be doing this!
+        // https://serde.rs/impl-serialize.html!
+        let arr = self.0.serialize();
+        let mut seq = serializer.serialize_tuple(arr.len())?;
+        for elem in &arr[..] {
+            seq.serialize_element(elem)?;
+        }
+        seq.end()
```

Comment:
I think once you have the `arr`, you can use its `serde::Serialize` implementation with e.g.:

``
```suggestion
        serde::Serialize::serialize(&arr, serializer)
```